model/storage/src/main/java/org/keycloak/storage/adapter/AbstractUserAdapterFederatedStorage.java

```diff
@@ -429,4 +430,19 @@ public Streams(final KeycloakSession session, final RealmModel realm, final Comp
             super(session, realm, storageProviderModel);
         }
     }
+
+    private MultivaluedHashMap<String, String> getFederatedUserAttributes() {
+        if (federatedAttributes == null) {
+            federatedAttributes = getFederatedStorage().getAttributes(realm, this.getId());
+        }
+        return Optional.ofNullable(federatedAttributes).map(MultivaluedHashMap::new).orElse(new MultivaluedHashMap<>());
+    }
+
+    private void setFederatedUserAttribute(String name, List<String> values) {
+        getFederatedStorage().setAttribute(realm, this.getId(), mapAttribute(name), values);
+        if (federatedAttributes == null) {
+            federatedAttributes = new MultivaluedHashMap<>();
+        }
+        federatedAttributes.put(name, values);
```

Comment:
This will IMHO lead to wrong results, as in the case of a previously `null` valued `federatedAttributes` it will initialize the map with one entry. The method above will then assume that this is the complete set of attributes and will ignore all existing attributes in the database. 

Instead, I'd suggest to not put anything into `federatedAttributes` if the map is currently `null`.

Reply:
What if we initialize the map with the results from `getFederatedUserAttributes`?

Reply:
Either that, or sipping the `put` if it is `null`.